compute_tools/src/bin/zenith_ctl.rs

```diff
@@ -0,0 +1,263 @@
+    // Usage of `pkill -P` or `kill -INT -zenith_ctl_pid` helps although.
+    let ecode = guard.child.wait().expect("failed to wait on postgres");
+
+    Ok(ecode)
+}
+
+fn main() -> Result<()> {
+    // TODO: re-use `zenith_utils::logging` later
+    init_logger(DEFAULT_LOG_LEVEL)?;
+
+    let matches = clap::App::new("zenith_ctl")
+        .version("0.1.0")
+        .arg(
+            clap::Arg::with_name("connstr")
+                .short("C")
+                .long("connstr")
+                .value_name("DATABASE_URL")
+                .required(true),
+        )
+        .arg(
+            clap::Arg::with_name("pgdata")
+                .short("D")
+                .long("pgdata")
+                .value_name("DATADIR")
+                .required(true),
+        )
+        .arg(
+            clap::Arg::with_name("pgbin")
+                .short("b")
+                .long("pgbin")
+                .value_name("POSTGRES_PATH"),
+        )
+        .arg(
+            clap::Arg::with_name("spec")
+                .short("s")
+                .long("spec")
+                .value_name("SPEC_JSON"),
+        )
+        .arg(
+            clap::Arg::with_name("spec-path")
+                .short("S")
+                .long("spec-path")
+                .value_name("SPEC_PATH"),
+        )
+        .get_matches();
+
+    let pgdata = matches.value_of("pgdata").expect("PGDATA path is required");
+    let connstr = matches
+        .value_of("connstr")
+        .expect("Postgres connection string is required");
+    let spec = matches.value_of("spec");
+    let spec_path = matches.value_of("spec-path");
+
+    // Try to use just 'postgres' if no path is provided
+    let pgbin = matches.value_of("pgbin").unwrap_or("postgres");
+
+    let spec: ClusterSpec = match spec {
+        // First, try to get cluster spec from the cli argument
+        Some(json) => serde_json::from_str(json)?,
+        None => {
+            // Second, try to read it from the file if path is provided
+            if let Some(sp) = spec_path {
+                let path = Path::new(sp);
+                let file = File::open(path)?;
+                serde_json::from_reader(file)?
+            } else {
+                // Finally, try to fetch it from the env
+                // XXX: not tested well and kept as a backup option for k8s, Docker, etc.
+                // TODO: remove later
+                match env::var("CLUSTER_SPEC") {
+                    Ok(json) => serde_json::from_str(&json)?,
+                    Err(_) => panic!("cluster spec should be provided via --spec, --spec-path or env variable CLUSTER_SPEC")
+                }
+            }
+        }
+    };
+
+    let compute_state = ComputeState {
+        connstr: connstr.to_string(),
+        pgdata: pgdata.to_string(),
+        pgbin: pgbin.to_string(),
+        spec,
+        ready: false,
+        last_active: Utc::now(),
+    };
+    let compute_state = Arc::new(RwLock::new(compute_state));
```

Comment:
You don't have to wrap the whole struct in RwLock, because only `Utc` can't be updated trivially. I'd introduce the following changes:

```
ready: AtomicBool,
last_active: RwLock<Utc>, // Or even AtomicU64
```

Reply:
My first intention was to use atomics here, but then I thought about:
- `ready` is likely to become a enum of at least (init, ready, running)
- Postgres child process handle is going to move into this struct, so we were able to update it's settings and send a SIGHUP signal, but we don't want to allow doing that concurrently
- We don't expect any real contention here

That way, it could be probably a combination of a couple of atomics + some sub-structure under mutex, but that just adds more complexity and I don't see any benefits now

Reply:
> I don't see any benefits now

The benefit is that when you see `&ComputeState`, it's immediately obvious what can be changed. Rust isn't always about what's easier to write, but it gives us tools to make our intentions clear.

Reply:
> The benefit is that when you see &ComputeState, it's immediately obvious what can be changed

Makes sense. Listed it for TODO here #1070